core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -0,0 +1,259 @@
+
+import java.util.AbstractMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Main class to swap the index pointed to by an alias, given some predicates
+ */
+public class TransportRolloverAction extends TransportMasterNodeAction<RolloverRequest, RolloverResponse> {
+
+    private final MetaDataCreateIndexService createIndexService;
+    private final MetaDataIndexAliasesService indexAliasesService;
+    private final Client client;
+
+    @Inject
+    public TransportRolloverAction(Settings settings, TransportService transportService, ClusterService clusterService,
+                                   ThreadPool threadPool, MetaDataCreateIndexService createIndexService,
+                                   ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                   MetaDataIndexAliasesService indexAliasesService, Client client) {
+        super(settings, RolloverAction.NAME, transportService, clusterService, threadPool, actionFilters, indexNameExpressionResolver,
+            RolloverRequest::new);
+        this.createIndexService = createIndexService;
+        this.indexAliasesService = indexAliasesService;
+        this.client = client;
+    }
+
+    @Override
+    protected String executor() {
+        // we go async right away
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected RolloverResponse newResponse() {
+        return new RolloverResponse();
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(RolloverRequest request, ClusterState state) {
+        IndicesOptions indicesOptions = IndicesOptions.fromOptions(true, true,
+            request.indicesOptions().expandWildcardsOpen(), request.indicesOptions().expandWildcardsClosed());
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE,
+            indexNameExpressionResolver.concreteIndexNames(state, indicesOptions, request.indices()));
+    }
+
+    @Override
+    protected void masterOperation(final RolloverRequest rolloverRequest, final ClusterState state,
+                                   final ActionListener<RolloverResponse> listener) {
+        final MetaData metaData = state.metaData();
+        validate(metaData, rolloverRequest);
+        final AliasOrIndex aliasOrIndex = metaData.getAliasAndIndexLookup().get(rolloverRequest.getSourceAlias());
+        final IndexMetaData indexMetaData = aliasOrIndex.getIndices().get(0);
+        final String sourceIndexName = indexMetaData.getIndex().getName();
+        client.admin().indices().prepareStats(sourceIndexName).clear().setDocs(true).execute(
+            new ActionListener<IndicesStatsResponse>() {
+                @Override
+                public void onResponse(IndicesStatsResponse indicesStatsResponse) {
+                    final IndexMetaData sourceIndex = metaData.index(sourceIndexName);
```

Comment:
can you move the calcualtions out into static methods so we can unittest them? Look at how I wrote `MetaDataCreateIndexServiceTests`